File: java/src/main/java/com/penapereira/patterns/templatemethod/AbstractClass.java
package com.penapereira.patterns.templatemethod;

/** Fixes the algorithm's skeleton; subclasses supply the steps. */
public abstract class AbstractClass {

    /** The skeleton: final so subclasses vary the steps, never the order. */
    public final String templateMethod() {
        return primitiveOperation1() + " then " + primitiveOperation2() + hook();
    }

    protected abstract String primitiveOperation1();

    protected abstract String primitiveOperation2();

    /** A hook: default behaviour that subclasses may extend. */
    protected String hook() {
        return "";
    }
}

File: python/src/patterns/templatemethod/abstract_class.py
from abc import ABC, abstractmethod


class AbstractClass(ABC):
    """Fixes the algorithm's skeleton; subclasses supply the steps."""

    def template_method(self) -> str:
        """The skeleton; subclasses vary the steps, never the order."""
        return f"{self.primitive_operation_1()} then {self.primitive_operation_2()}{self.hook()}"

    @abstractmethod
    def primitive_operation_1(self) -> str: ...

    @abstractmethod
    def primitive_operation_2(self) -> str: ...

    def hook(self) -> str:
        """A hook: default behaviour that subclasses may extend."""
        return ""
